### ISSUE TITLE
Thermal dynamics tick handlers are taking up a majority of CPU time, and seem to increase slowly over time

### ISSUE THREAD

**Issue by Ricket** [CoFH/Feedback#395]
I am running the ATM2 pack v0.73, which has ThermalDynamics-1.11.2-2.2.5.12.

We have seen that server CPU usage creeps up over time, eventually it is stuck around 100% (according to top; so that's about 100% of one core) and we start seeing consistent "can't keep up" messages. We also, after the server is in this state, have issues where we try to log in and then time out. I'm not sure if that is the same issue or a different issue, but it does not happen if we log in shortly after starting the server.

I am trying to profile and find the cause of the issues, and in doing so I found that Thermal Dynamics is responsible for the vast majority of the CPU usage of the main Minecraft thread (the other threads are sleeping or blocking on read).

![Screenshot from YourKit profile session](https://i.imgur.com/rKJOobM.png)

The main thread is basically not sleeping at this point (thus the 'can't keep up').

I play with one other person in this server that I maintain, and while I only have a single Thermal machine and no piping, he has been using the piping, including fluid pipes, which seems to be the primary culprit. I'm sure his creations are solely responsible for the time taken by the Thermal tick methods. But I asked him if he had an excessive amount and he said it was definitely not excessive. I can get more details than that if it will help.

The other thing I'm curious about: I took the sample above from a server which I merely started and nobody logged into. This was after the server was running for about 18 hours. There are no chunkloaders in the world. What is thermal doing, why is it doing any calculations at all? Are the chunks being chunkloaded even though we didn't put down a chunkloader?

I'm glad to provide more debug info. I haven't yet dug into the ThermalDynamics code, besides just browsing some of these at-fault methods and not seeing an obvious cause. I might do so.

**Comment by KingLemming** [comment]
This isn't something we've seen at all, and we stress tested with literally millions of ducts.

**Comment by KingLemming** [comment]
Yeah, we're gonna need loads more parsing, logs, and whatever else you got. I see Mekanism in that log and tbh that automatically makes me a bit worried.

**Comment by Ricket** [comment]
Right. He has a simple mekanism 5x ore processing chain and since Mekanism pipes are broken, he is using thermal pipes between everything (for the fluids, items, and energy).

When I drilled deeper into the time usage, Mekanism was taking up a majority of the time with its calculations. But I have to wonder if it's sort of a joint issue between the two mods. My general thinking is, why is thermal doing these calculations every tick when the chunks are unloaded (at least we assume they are, since nobody is logged in and there are no chunkloaders), but simultaneously, Mekanism is probably not written efficiently for these areas of code.

I'm not sure it would be safe to upload a heap dump, but I can definitely give you the values of any variables if you are curious about them (the length of particular lists or maps for example). The log doesn't really say anything, which is why I had to look into profiling.

And I will give more info about the setup. I'll try to make a repro setup using only mekanism and thermal, nothing else. Like I said, I don't think it's only mek or only thermal at this point. But if it is only one of those then I do believe at this time that mek is most at fault.

Let me know if there's any variable values you want, and when I get a chance (this weekend hopefully) I will test things and give a detailed info about the repro setup.

**Comment by KingLemming** [comment]
Right, had a feeling.

The problem with transport mods is that they sort of get hit with any inefficiencies with anything they interact with.

There's no joint issue here - Mekanism just has bad code. And Thermal is doing the calculations because the chunks *aren't* unloaded - there's a good chance Mekanism is force loading them. These are multiblocks, right?

**Comment by Ricket** [comment]
I've been working on this for the past several hours. I made a custom build of Mekanism which optimizes the methods in the profile results (energy accept/draw, fluid fill), making them perform a lot better. Will put out pull request(s) to the project. I believe this _helps_ the issue.

I can't yet explain the issue where it gets worse over time. Will hopefully do some more debugging tomorrow. I have also not pinpointed whether chunks are being loaded or not. I hope that Thermal is not doing any processing when a grid is fully unloaded. I also hope that the processing that it does is optimal, for example if no consumers take anything and the grid is full then don't attempt to pull from producers.

Let's please keep this issue open and I'll continue to update. For now I need to go prepare the pull request for Mekanism.

**Comment by KingLemming** [comment]
Thermal does not do ANYTHING if it isn't loaded. That's literally how it works. So I'm happy to leave this up, but truly - the issue isn't on TD's side.

The ducts are at the mercy of the things they are attached to. So for example, if a tile entity outright lies about something, the duct has no choice but to believe it. A badly coded tile entity (such as Railcraft's void chest) can definitely hurt TPS.

**Comment by Ricket** [comment]
My server's been running ok so I have calmed down on the debugging, (I'm no longer getting any "can't keep up" messages). The latest Mekanism has my changes in it, and it has much better performance when it comes to the energy calculations.

I took a couple memory dumps and in every case, the ForgeChunkLoader had no tickets in its map, which should indicate no chunks loaded. In the Thermal Dynamics space, it showed around 1500 grids which seemed too high to me, but I can't count them all by hand so I don't know. I did see times where nobody was logged in and there was no Thermal Dynamics activity as in my above screenshot. But as I reported originally, I definitely have times where nobody is logged in and the TickHandler is processing everything. I haven't found a pattern to it.

I created a modpack with only thermal and was not able to reproduce a leak in single player mode. (I did get a couple false positives, but I learned that visualvm's sampler also shows objects that are waiting to be gc'd) I'd like to do more testing in a server-client situation and logging in and out. Maybe also analyze some more memory dumps of my server.

**Comment by KingLemming** [comment]
Closing this down. We did a LOT of parsing and haven't been able to identify any situations where this is actually an issue caused by us.

**Comment by douglasg14b** [comment]
I've noticed this when I'm trying to push stacks to very fast machines as fast as possible and items are bouncing back and forth between the machine and the servo because there are constant misses.

I can hook 6pipe sup with resonant servos and let it try and fill something that only hold 1 stack, but empties every 1/5th of a second or so, and the tick time doubles. 

It's anecdotal, but it's what I managed to find.
